libs/trustchain/src/qrcode/index.ts

```diff
@@ -0,0 +1,258 @@
+import { crypto } from "@ledgerhq/hw-trustchain";
+import { createCustomErrorClass } from "@ledgerhq/errors";
+import { getEnv } from "@ledgerhq/live-env";
+import WebSocket from "ws";
+import { LiveCredentials } from "../types";
+
+export const InvalidDigitError = createCustomErrorClass("InvalidDigitError");
+
+const version = 1;
+
+/**
+ * establish a channel to be able to add a member to the trustchain after displaying the QR Code
+ * @returns a promise that resolves when this is done
+ */
+export async function createQRCodeHostInstance({
+  onDisplayQRCode,
+  onDisplayDigits,
+  addMember,
+}: {
+  /**
+   * this function will need to display a UI to show the QR Code
+   */
+  onDisplayQRCode: (url: string) => void;
+  /**
+   * this function will need to display a UI to show the digits
+   */
+  onDisplayDigits: (digits: string) => void;
+  /**
+   * this function will need to using the TrustchainSDK (and use sdk.addMember)
+   */
+  addMember: (publicKey: string) => Promise<void>;
+}): Promise<void> {
+  const ephemeralKey = await crypto.randomKeypair();
+  const publisher = crypto.to_hex(ephemeralKey.publicKey);
+  let sessionEncryptionKey: Uint8Array | undefined;
+  let cipher: ReturnType<typeof makeCypher> | undefined;
+  let expectedDigits: string | undefined;
+  const url = `${getEnv("TRUSTCHAIN_API")}/v1/qr?host=${publisher}`;
+  onDisplayQRCode(url);
+  const ws = new WebSocket(url);
+  function send(type: string, payload: object) {
+    ws.send(
+      JSON.stringify({
+        version,
+        publisher,
+        message: type,
+        payload,
+      }),
+    );
+  }
+
+  return new Promise((resolve, reject) => {
+    ws.on("error", reject);
+    ws.on("message", async e => {
+      try {
+        const data = parseMessage(e);
+        switch (data.message) {
+          case "InitiateHandshake": {
+            const candidatePublicKey = crypto.from_hex(data.publisher);
+            sessionEncryptionKey = await crypto.ecdh(ephemeralKey, candidatePublicKey);
+            cipher = makeCypher(sessionEncryptionKey);
+            // --- end of handshake first phase ---
+            const digitsCount = 3;
+            const digits = await randomDigits(digitsCount);
+            expectedDigits = digits;
+            onDisplayDigits(digits);
+            send("HandshakeChallenge", { digits: digitsCount, connected: false });
+            break;
+          }
+          case "CompleteHandshakeChallenge": {
+            if (!cipher) {
+              throw new Error("sessionEncryptionKey not set");
+            }
+            const { encrypted } = data.payload as { encrypted: string };
+            const { digits } = (await cipher.decrypt(encrypted)) as { digits: string };
+            if (digits !== expectedDigits) {
+              // FIXME: should we send failure to the backend?
```

Comment:
Indeed, you can send this message:
```
{
        "version": 1,
        "publisher": XXXX,
        "message": "Failure",
        "payload": {
          "type": "HANDSHAKE_COMPLETION_FAILED",
          "message": "Some message that could be displayed to end user"
        }
      }
```
=> the server will forward it to the candidate, and then disconnect both.

Alternative is to just disconnect => the server will also disconnect the candidate, so it knows it failed the challenge implictly.

I suggest going for the more explicit solution, in case we want to allow retries in the future (currently, server is very strict: any failure = broadcast the error message and drop both clients)